// nft.storage/dist/src/platform.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var fetch = require('@web-std/fetch');
var fetch__default = _interopDefault(fetch);
var formData = require('@web-std/form-data');
var blob = require('@web-std/blob');
var file = require('@web-std/file');



Object.defineProperty(exports, 'Headers', {
  enumerable: true,
  get: function () {
    return fetch.Headers;
  }
});
Object.defineProperty(exports, 'Request', {
  enumerable: true,
  get: function () {
    return fetch.Request;
  }
});
Object.defineProperty(exports, 'Response', {
  enumerable: true,
  get: function () {
    return fetch.Response;
  }
});
exports.fetch = fetch__default;
Object.defineProperty(exports, 'FormData', {
  enumerable: true,
  get: function () {
    return formData.FormData;
  }
});
Object.defineProperty(exports, 'Blob', {
  enumerable: true,
  get: function () {
    return blob.Blob;
  }
});
Object.defineProperty(exports, 'ReadableStream', {
  enumerable: true,
  get: function () {
    return blob.ReadableStream;
  }
});
Object.defineProperty(exports, 'File', {
  enumerable: true,
  get: function () {
    return file.File;
  }
});
